code/modules/projectiles/projectile.dm

```diff
@@ -293,211 +260,320 @@
+
+/obj/item/projectile/proc/process_hitscan()
+	var/safety = range * 3
+	record_hitscan_start(RETURN_POINT_VECTOR_INCREMENT(src, Angle, MUZZLE_EFFECT_PIXEL_INCREMENT, 1))
+	while(loc && !QDELETED(src))
+		if(paused)
+			stoplag(1)
+			continue
+		if(safety-- <= 0)
+			qdel(src)
+			crash_with("WARNING: [type] projectile encountered infinite recursion during hitscanning in [__FILE__]/[__LINE__]!")
+			return	//Kill!
+		pixel_move(1, 1, TRUE)
+
+/obj/item/projectile/proc/record_hitscan_start(datum/point/pcache)
+	beam_segments = list()	//initialize segment list with the list for the first segment
+	beam_index = pcache
+	beam_segments[beam_index] = null	//record start.
+
+/obj/item/projectile/proc/vol_by_damage()
+	if(src.damage)
+		return Clamp((src.damage) * 0.67, 30, 100)// Multiply projectile damage by 0.67, then CLAMP the value between 30 and 100
+	else
+		return 50 //if the projectile doesn't do damage, play its hitsound at 50% volume.
 
-	if(ispath(tracer_type))
-		var/obj/effect/projectile/P = new tracer_type(location.loc)
+/obj/item/projectile/proc/before_z_change(turf/oldloc, turf/newloc)
+	var/datum/point/pcache = trajectory.copy_to()
+	if(hitscan)
+		store_hitscan_collision(pcache)
 
-		if(istype(P))
-			P.set_transform(M)
-			P.pixel_x = location.pixel_x
-			P.pixel_y = location.pixel_y
-			if(!hitscan)
-				P.activate(step_delay)	//if not a hitscan projectile, remove after a single delay
-			else
-				P.activate()
-
-/obj/item/projectile/proc/impact_effect(var/matrix/M)
-	if (!location)
-		return//Combat drones sometimes cause a runtime error with null location. Impact effect isnt important
-	if(ispath(tracer_type))
-
-		var/obj/effect/projectile/P = new impact_type(location.loc)
-
-		if(istype(P))
-			P.set_transform(M)
-			P.pixel_x = location.pixel_x
-			P.pixel_y = location.pixel_y
-			P.activate()
-
-//"Tracing" projectile
-/obj/item/projectile/test //Used to see if you can hit them.
-	invisibility = 101 //Nope!  Can't see me!
-	yo = null
-	xo = null
-	var/result = 0 //To pass the message back to the gun.
-
-/obj/item/projectile/test/Bump(atom/A as mob|obj|turf|area)
-	if(A == firer)
-		loc = A.loc
-		return //cannot shoot yourself
-	if(istype(A, /obj/item/projectile))
+/obj/item/projectile/process()
+	last_process = world.time
+	if(!loc || !fired || !trajectory)
+		fired = FALSE
+		return PROCESS_KILL
+	if(paused || !isturf(loc))
+		last_projectile_move += world.time - last_process		//Compensates for pausing, so it doesn't become a hitscan projectile when unpaused from charged up ticks.
 		return
-	if(istype(A, /mob/living) || istype(A, /obj/mecha) || istype(A, /obj/vehicle))
-		result = 2 //We hit someone, return 1!
+	var/elapsed_time_deciseconds = (world.time - last_projectile_move) + time_offset
+	time_offset = 0
+	var/required_moves = speed > 0? Floor(elapsed_time_deciseconds / speed, 1) : MOVES_HITSCAN			//Would be better if a 0 speed made hitscan but everyone hates those so I can't make it a universal system :<

```

Comment:
Any benefit to implementing this comment?

Reply:
Woops leftover from tg where I had to do simulated hitscan due to lack of true hitscan when this was first made.